parquet-variant/src/main/java/org/apache/parquet/variant/VariantUtil.java

```diff
@@ -0,0 +1,659 @@
+        }
+        result = new BigDecimal(new BigInteger(bytes), scale);
+        break;
+      default:
+        throw unexpectedType(
+            new Variant.Type[] {Variant.Type.DECIMAL4, Variant.Type.DECIMAL8, Variant.Type.DECIMAL16});
+    }
+    return result;
+  }
+
+  static BigDecimal getDecimal(ByteBuffer value) {
+    return getDecimalWithOriginalScale(value);
+  }
+
+  static float getFloat(ByteBuffer value) {
+    checkIndex(value.position(), value.limit());
+    int basicType = value.get(value.position()) & BASIC_TYPE_MASK;
+    int typeInfo = (value.get(value.position()) >> BASIC_TYPE_BITS) & PRIMITIVE_TYPE_MASK;
+    if (basicType != PRIMITIVE || typeInfo != FLOAT) {
+      throw unexpectedType(Variant.Type.FLOAT);
+    }
+    return Float.intBitsToFloat((int) readLong(value, value.position() + 1, 4));
+  }
+
+  static ByteBuffer getBinary(ByteBuffer value) {
+    checkIndex(value.position(), value.limit());
+    int basicType = value.get(value.position()) & BASIC_TYPE_MASK;
+    int typeInfo = (value.get(value.position()) >> BASIC_TYPE_BITS) & PRIMITIVE_TYPE_MASK;
+    if (basicType != PRIMITIVE || typeInfo != BINARY) {
+      throw unexpectedType(Variant.Type.BINARY);
+    }
+    int start = value.position() + 1 + U32_SIZE;
+    int length = readUnsigned(value, value.position() + 1, U32_SIZE);
+    checkIndex(start + length - 1, value.limit());
+    return slice(value, start);
+  }
+
+  static String getString(ByteBuffer value) {
+    checkIndex(value.position(), value.limit());
+    int basicType = value.get(value.position()) & BASIC_TYPE_MASK;
+    int typeInfo = (value.get(value.position()) >> BASIC_TYPE_BITS) & PRIMITIVE_TYPE_MASK;
+    if (basicType == SHORT_STR || (basicType == PRIMITIVE && typeInfo == LONG_STR)) {
+      int start;
+      int length;
+      if (basicType == SHORT_STR) {
+        start = value.position() + 1;
+        length = typeInfo;
+      } else {
+        start = value.position() + 1 + U32_SIZE;
+        length = readUnsigned(value, value.position() + 1, U32_SIZE);
+      }
+      checkIndex(start + length - 1, value.limit());
+      if (value.hasArray()) {
+        // If the buffer is backed by an array, we can use the array directly.
+        return new String(value.array(), value.arrayOffset() + start, length);
+      } else {
+        // If the buffer is not backed by an array, we need to copy the bytes into a new array.
+        byte[] valueArray = new byte[length];
+        slice(value, start).get(valueArray);
+        return new String(valueArray);
+      }
+    }
+    throw unexpectedType(Variant.Type.STRING);
+  }
+
+  static java.util.UUID getUUID(ByteBuffer value) {
+    checkIndex(value.position(), value.limit());
+    int basicType = value.get(value.position()) & BASIC_TYPE_MASK;
+    int typeInfo = (value.get(value.position()) >> BASIC_TYPE_BITS) & PRIMITIVE_TYPE_MASK;
+    if (basicType != PRIMITIVE || typeInfo != UUID) {
+      throw unexpectedType(Variant.Type.UUID);
```

Comment:
Minor: it would be nice to include what the type actual was.

Is there a guarantee that the type is always checked when reading through these util methods?